scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,581 @@
+    wj[0] = wj[0] / 2 if k == 0 else wj[0]
+
+    return xjc, wj
+
+
+def _pair_cache(max_level, h0):
+    # Cache the ascissa-weight pairs up to a specified level
+    # All abscissae (weights) are stored concatenated in a 1D array;
+    # `index` notes the last index of each level.
+    pairs = [_compute_pair(k, h0) for k in range(max_level + 1)]
+    lengths = [len(pair[0]) for pair in pairs]
+    _pair_cache.xjc, _pair_cache.wj = np.concatenate(pairs, axis=-1)
+    _pair_cache.indices = np.cumsum(lengths)
+_pair_cache.xjc = None
+_pair_cache.wj = None
+_pair_cache.indices = []
+
+
+def _get_pairs(k, h0, inclusive=False):
+    # Retrieve the specified abscissa-weight pairs from the cache
+    # If `inclusive`, return all up to and including the specified level
+    indices = _pair_cache.indices
+    if len(indices) <= k:
+        _pair_cache(max(10, k), h0)
+        indices = _pair_cache.indices
+    xjc = _pair_cache.xjc
+    wj = _pair_cache.wj
+
+    start = 0 if (k == 0 or inclusive) else indices[k-1]
+    end = indices[k]
+
+    return xjc[start:end].copy(), wj[start:end].copy()
+
+
+def _transform_to_limits(a, b, xjc, wj):
+    # Transform integral according to user-specified limits. This is just
+    # math that follows from the fact that the standard limits are (-1, 1).
+    # Note: If we had stored xj instead of xjc, we would have
+    # xj = alpha * xj + beta, where beta = (a + b)/2
+    alpha = (b - a) / 2
+    xj = np.concatenate((-alpha * xjc + b, alpha * xjc + a), axis=-1)
+    wj = wj*alpha  # these need to get broadcasted, so no *=
+    wj = np.concatenate((wj, wj), axis=-1)
+
+    # Points at the boundaries can be generated due to finite precision
+    # arithmetic, but these function values aren't supposed to be included in
+    # the Euler-Maclaurin sum. Ideally we wouldn't evaluate the function at
+    # these points; however, we can't easily filter out points when this
+    # function is vectorized. Instead, zero the weights.
+    invalid = (xj <= a) | (xj >= b)
+    wj[invalid] = 0
+    return xj, wj
+
+
+def _euler_maclaurin_sum(f, h, xj, wj, last_terms, log):
+    # Perform the Euler-Maclaurin Sum, [1] Section 4
+    with np.errstate(divide='ignore', over='ignore', invalid='ignore'):
+        # Warnings due to function evaluation at endpoints are not cause for
+        # concern; the resulting values will be given 0 weight.
+        fj = f(xj)
+
+    # The error estimate needs to know the magnitude of the last term
+    # omitted from the Euler-Maclaurin sum. This is a bit involved because
+    # it may have been computed at a previous level. I sure hope it's worth
+    # all the trouble.
+    xl0, fl0, wr0, xr0, fr0, wl0, d4 = last_terms  # incumbent last terms
+
+    # Find the most extreme abscissae corresponding with terms that are
+    # included in the sum in this level.
+    invalid = ~np.isfinite(fj) | (wj == 0)  # these terms aren't in the sum
```

Comment:
This `|` and the zero weights for abscissae on the boundary may need to be reconsidered. Infinities on the boundary, infinities not on the boundary, and finite numbers on the boundary are different. Certainly interior infinities should not be lumped in here, and we may be able to take shortcuts when there are no infinities. (For example, if there is no infinity at the boundary, we don't *need* to exclude any terms from the sum, and then we know the index containing the rightmost point.)

Also, shouldn't `d4` include a factor of `h`? Don't we need to track `d3` across multiple levels, too?